src/block_cache/block_cache.c

```diff
@@ -505,21 +505,32 @@ __wt_blkcache_put(
  * __wt_blkcache_remove --
  *     Remove a block from the cache.
  */
-void
+int
 __wt_blkcache_remove(WT_SESSION_IMPL *session, const uint8_t *addr, size_t addr_size)
 {
     WT_BLKCACHE *blkcache;
     WT_BLKCACHE_ITEM *blkcache_item;
-    uint64_t bucket, hash;
+    uint64_t bucket, hash, sleep_usecs, yield_count;
 
     blkcache = &S2C(session)->blkcache;
-
     hash = __wt_hash_city64(addr, addr_size);
     bucket = hash % blkcache->hash_size;
+    sleep_usecs = yield_count = 0;
+
     __wt_spin_lock(session, &blkcache->hash_locks[bucket]);
     TAILQ_FOREACH (blkcache_item, &blkcache->hash[bucket], hashq) {
         if (blkcache_item->addr_size == addr_size && blkcache_item->fid == S2BT(session)->id &&
           memcmp(blkcache_item->addr, addr, addr_size) == 0) {
+            /*
+             * The block might be in use by another thread, wait for it to be released before
+             * freeing it.
+             */
+            while (blkcache_item->ref_count != 0) {
+                __wt_spin_backoff(&yield_count, &sleep_usecs);
+                if (++yield_count > WT_THOUSAND * 10)
+                    return (EBUSY);
+                WT_STAT_CONN_INCRV(session, block_cache_blocks_removed_blocked, sleep_usecs);
+            }
             TAILQ_REMOVE(&blkcache->hash[bucket], blkcache_item, hashq);
             __blkcache_update_ref_histogram(session, blkcache_item, BLKCACHE_RM_FREE);
             __wt_spin_unlock(session, &blkcache->hash_locks[bucket]);
```

Comment:
I think you want to move the spinning here (i.e. to replace the assertion). You want to do it after removing it from the list and after unlocking. We don't want to sleep while holding the lock. That will prevent every block in that bucket from being accessed and could be a performance hit. 

By removing while holding the lock, it prevents a thread racing in `wt_blkcache_get` from getting the lock and finding this block at this point. Now the new code will allow the removal thread to wait for any thread that had it before this call to complete its usage before the removal frees it.

Reply:
Sounds good, I made the changed here 4d12c7313e601e312b2065089ffa1bb80099ecb2. I also took into account Monica's suggestion where we would do the spinning after incrementing all the stats, are you ok with this too?